File: leetcode/0255-verify-preorder-sequence-in-binary-search-tree/Solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public boolean verifyPreorder(int[] preorder) {
        Deque<Integer> stack = new ArrayDeque<>();
        long lowerBound = Long.MIN_VALUE;

        for (int num : preorder) {
            if (num < lowerBound) {
                return false;
            }
            while (!stack.isEmpty() && stack.peek() < num) {
                lowerBound = stack.pop();
            }
            stack.push(num);
        }

        return true;
    }
}

File: leetcode/0255-verify-preorder-sequence-in-binary-search-tree/solution.cpp
#include <vector>
#include <stack>
#include <climits>
using namespace std;

class Solution {
public:
    bool verifyPreorder(vector<int>& preorder) {
        stack<int> st;
        long long lowerBound = LLONG_MIN;

        for (int num : preorder) {
            if (num < lowerBound) {
                return false;
            }
            while (!st.empty() && st.top() < num) {
                lowerBound = st.top();
                st.pop();
            }
            st.push(num);
        }

        return true;
    }
};
